src/tdycharacteristiccurves.c

```diff
@@ -0,0 +1,283 @@
+#include <private/tdycoreimpl.h>
+#include <private/tdycharacteristiccurvesimpl.h>
+#include <private/tdymemoryimpl.h>
+
+PetscErrorCode CharacteristicCurveCreate(PetscInt ncells, CharacteristicCurve **_cc){
+
+  PetscFunctionBegin;
+  PetscErrorCode ierr;
+
+  *_cc = (CharacteristicCurve *)malloc(sizeof(struct _CharacteristicCurve));
+
+  ierr = PetscMalloc(ncells*sizeof(PetscInt),&((*_cc)->SatFuncType)); CHKERRQ(ierr);
+  ierr = PetscMalloc(ncells*sizeof(PetscInt),&(*_cc)->RelPermFuncType); CHKERRQ(ierr);
+
+  ierr = PetscMalloc(ncells*sizeof(PetscReal),&(*_cc)->Kr); CHKERRQ(ierr);
+  ierr = PetscMalloc(ncells*sizeof(PetscReal),&(*_cc)->dKr_dS); CHKERRQ(ierr);
+  ierr = PetscMalloc(ncells*sizeof(PetscReal),&(*_cc)->S); CHKERRQ(ierr);
+  ierr = PetscMalloc(ncells*sizeof(PetscReal),&(*_cc)->dS_dP); CHKERRQ(ierr);
+  ierr = PetscMalloc(ncells*sizeof(PetscReal),&(*_cc)->d2S_dP2); CHKERRQ(ierr);
+  ierr = PetscMalloc(ncells*sizeof(PetscReal),&(*_cc)->dS_dT); CHKERRQ(ierr);
+  ierr = PetscMalloc(ncells*sizeof(PetscReal),&(*_cc)->sr); CHKERRQ(ierr);
+  ierr = PetscMalloc(ncells*sizeof(PetscReal),&(*_cc)->m); CHKERRQ(ierr);
+  ierr = PetscMalloc(ncells*sizeof(PetscReal),&(*_cc)->n); CHKERRQ(ierr);
+  ierr = PetscMalloc(ncells*sizeof(PetscReal),&(*_cc)->alpha); CHKERRQ(ierr);
+  
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode CharacteristicCurveDestroy(CharacteristicCurve *cc){
+
+  PetscFunctionBegin;
+  PetscErrorCode ierr;
+
+  if (cc->SatFuncType    ) { ierr = PetscFree(cc->SatFuncType    ); CHKERRQ(ierr); }
+  if (cc->RelPermFuncType) { ierr = PetscFree(cc->RelPermFuncType); CHKERRQ(ierr); }
+  if (cc->Kr             ) { ierr = PetscFree(cc->Kr             ); CHKERRQ(ierr); }
+  if (cc->dKr_dS         ) { ierr = PetscFree(cc->dKr_dS         ); CHKERRQ(ierr); }
+  if (cc->S              ) { ierr = PetscFree(cc->S              ); CHKERRQ(ierr); }
+  if (cc->dS_dP          ) { ierr = PetscFree(cc->dS_dP          ); CHKERRQ(ierr); }
+  if (cc->d2S_dP2        ) { ierr = PetscFree(cc->d2S_dP2        ); CHKERRQ(ierr); }
+  if (cc->dS_dT          ) { ierr = PetscFree(cc->dS_dT          ); CHKERRQ(ierr); }
+  if (cc->sr             ) { ierr = PetscFree(cc->sr             ); CHKERRQ(ierr); }
+  if (cc->m              ) { ierr = PetscFree(cc->m              ); CHKERRQ(ierr); }
+  if (cc->n              ) { ierr = PetscFree(cc->n              ); CHKERRQ(ierr); }
+  if (cc->alpha          ) { ierr = PetscFree(cc->alpha          ); CHKERRQ(ierr); }
+  
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode TDySetResidualSaturationValuesLocal(TDy tdy, PetscInt ni, const PetscInt ix[], const PetscScalar y[]){
```

Comment:
FYI: In these kinds of functions, you can actually place `ni` within the square brackets for the `ix` and `y` arguments, to tell the compiler to expect arrays of this size.